paimon-core/src/main/java/org/apache/paimon/catalog/CachingCatalog.java

```diff
@@ -227,6 +244,47 @@ private void putTableCache(Identifier identifier, Table table) {
         tableCache.put(identifier, table);
     }
 
+    public List<PartitionEntry> getPartitions(Identifier identifier) throws TableNotExistException {
+        Table table = this.getTable(identifier);
+        if (enablePartitionCache(table)) {
+            List<PartitionEntry> partitions;
+            partitions = partitionCache.getIfPresent(identifier);
+            if (partitions == null || partitions.isEmpty()) {
+                partitions = this.refreshPartitions(identifier);
+            }
+            return partitions;
+        }
+        return ((FileStoreTable) table).newSnapshotReader().partitionEntries();
+    }
+
+    public List<PartitionEntry> refreshPartitions(Identifier identifier)
+            throws TableNotExistException {
+        Table table = this.getTable(identifier);
+        List<PartitionEntry> partitions =
+                ((FileStoreTable) table).newSnapshotReader().partitionEntries();
+        if (enablePartitionCache(table)
+                && partitionCache.asMap().values().stream().mapToInt(List::size).sum()
+                        < this.partitionMaxNum) {
+            partitionCache.put(identifier, partitions);
+        }
+        return partitions;
+    }
+
+    private boolean enablePartitionCache(Table table) {
+        return partitionCache != null
+                && table instanceof FileStoreTable
+                && !table.partitionKeys().isEmpty();
+    }
+
+    @Override
+    public void dropPartition(Identifier identifier, Map<String, String> partitions)
+            throws TableNotExistException, PartitionNotExistException {
+        wrapped.dropPartition(identifier, partitions);
+        if (partitionCache != null) {
+            partitionCache.invalidate(identifier);
```

Comment:
Why invalidate all partitions of this identifier in the cache?

Reply:
Can't find a better way, any ideas?